support-workers/src/main/scala/com/gu/emailservices/DigitalPackEmailFields.scala

```diff
@@ -1,78 +1,242 @@
+
+  implicit val encoder: Encoder.AsObject[DigitalSubscriptionEmailAttributes] = Encoder.AsObject.instance {
+    case v: BasicDSAttributes => v.asJsonObject
+    case v: DirectDSAttributes => v.asJsonObject
+    case v: GifteeRedemptionAttributes => v.asJsonObject
+    case v: GifterPurchaseAttributes => v.asJsonObject
+    case v: GifteeNotificationAttributes => v.asJsonObject
+  }
+
+
+  sealed trait PaymentFieldsAttributes extends Product with Serializable {
+    val default_payment_method: String
+  }
+  object PaymentFieldsAttributes {
+
+    implicit val e1: Encoder.AsObject[DDAttributes] = deriveEncoder
+    implicit val e2: Encoder.AsObject[CCAttributes] = deriveEncoder
+    implicit val e3: Encoder.AsObject[PPAttributes] = deriveEncoder
+
+    implicit val encoder: Encoder.AsObject[PaymentFieldsAttributes] = Encoder.AsObject.instance {
+      case v: DDAttributes => v.asJsonObject
+      case v: CCAttributes => v.asJsonObject
+      case v: PPAttributes => v.asJsonObject
+    }
+
+    case class DDAttributes(
+      account_number: String,
+      sort_code: String,
+      account_name: String,
+      mandateid: String,
+      default_payment_method: String = "Direct Debit",
+    ) extends PaymentFieldsAttributes
+
+    case class CCAttributes(
+      default_payment_method: String = "Credit/Debit Card",
+    ) extends PaymentFieldsAttributes
+
+    case class PPAttributes(
+      default_payment_method: String = "PayPal",
+    ) extends PaymentFieldsAttributes
+
+  }
+
+
+  case class BasicDSAttributes(
+    zuorasubscriberid: String,
+    emailaddress: String,
+    first_name: String,
+    last_name: String,
+    subscription_details: String,
+  ) extends DigitalSubscriptionEmailAttributes
+
+  case class DirectDSAttributes(
+    directCorp: BasicDSAttributes,
+    subscription_term: String,
+    payment_amount: String,
+    country: String,
+    date_of_first_payment: String,
+    currency: String,
+    trial_period: String,
+    paymentFieldsAttributes: PaymentFieldsAttributes,
+  ) extends DigitalSubscriptionEmailAttributes
+
+  case class GifteeRedemptionAttributes(
+    gift_recipient_first_name: String,
+    subscription_details: String,
+    gift_start_date: String,
+    gift_recipient_email: String,
+  ) extends DigitalSubscriptionEmailAttributes
+
+  case class GifterPurchaseAttributes(
+    gifter_first_name: String,
+    gifter_last_name: String,
+    gift_recipient_first_name: String,
+    gift_recipient_last_name: String,
+    gift_recipient_email: String,
+    gift_personal_message: String,
+    gift_code: String,
+    gift_delivery_date: String,
+    subscription_details: String,
+    paymentAttributes: PaymentFieldsAttributes,
+    date_of_first_payment: String,
+  ) extends DigitalSubscriptionEmailAttributes
+
+  case class GifteeNotificationAttributes(
+    gifter_first_name: String,
+    gift_personal_message: String,
+    gift_code: String,
+  ) extends DigitalSubscriptionEmailAttributes
```

Comment:
these lines define the interface between support-workers and braze emails in one place.  This should make it easier in future when working out the contract between us and braze.

Reply:
I'm also not that sure that all of what we do send is used by the emails so it would be nice to check that before writing a test

Reply:
agree it would be nice to reassess what is used.  a test would still provide value for regressions until then.

I wouldn't want to remove anything in this refactoring PR as if something breaks it adds uncertainty - in future either @GHaberis could check in braze, or we could do some eyeballing on the emails.  I'm not sure if some of the fields may be stored in braze for analysis purposes?  I think George was creating a doc for the braze->support workers interface but I'm not sure how far he got.

Reply:
I have added tests for my own piece of mind - they have shown that there are no changes to the JSON output for corporate and direct digital subs.  From visual looking, there are 3 fields we may be able to remove from DS direct so I may play with that a bit.
![image](https://user-images.githubusercontent.com/7304387/94176214-ae85bc80-fe8f-11ea-9a9e-c2c670a2cb49.png)